File: chapter1/001/main.go
package main

import (
	"fmt"
	"unicode/utf8"
)

func main() {
	fmt.Println(hasUniqueWithMap("abc"))
	fmt.Println(hasUniqueWithMap("stẽnio"))
	fmt.Println(hasUniqueWithMap("abba"))

	fmt.Println(hasUniqueInPlace("abc"))
	fmt.Println(hasUniqueInPlace("stẽnio"))
	fmt.Println(hasUniqueInPlace("abba"))
}

func hasUniqueWithMap(s string) bool {
	runes := make(map[rune]bool)

	for _, r := range s {
		if runes[r] {
			return false
		}
		runes[r] = true
	}

	return true
}

func hasUniqueInPlace(s string) bool {
	for i := 0; i < len(s); {
		r, size := utf8.DecodeRuneInString(s[i:])
		for j := i + size; j < len(s); {
			ra, sizeJ := utf8.DecodeRuneInString(s[j:])
			if r == ra {
				return false
			}
			j += sizeJ
		}
		i += size
	}
	return true
}

File: chapter1/001/main.py
def unique_map(string):
    m = {}
    for c in string:
        if m.get(c):
            return False
        m[c] = True
    return True


def unique_inplace(string):
    for i in range(len(string)):
        for j in range(i+1, len(string)):
            if string[i] == string[j]:
                return False
    return True


print(unique_map("abc"))
print(unique_map("stẽnio"))
print(unique_map("abba"))

print(unique_inplace("abc"))
print(unique_inplace("stẽnio"))
print(unique_inplace("abba"))